### PR TITLE
Temporary fix for the Julia GC.

### PR DESCRIPTION
**Pull request by rbehrends**
This addresses a problem where the Julia GC during a partial sweep frees
some, but not all objects of an unreachable data structure. If an
address to a remaining object of such a data structure is still randomly
hit during conservative stack scanning, it may erroneously try to mark
the deallocated objects, too.

This temporary fix works by validating each pointer before it is marked.
Because this is potentially expensive, the eventual solution should fix
the root cause, which is conservative stack scanning resurrecting
pointers to dead objects that have been freed during a partial
collection.

### CODE REVIEW

File: src/julia_gc.c [hunk 1]
```diff
@@ -54,6 +54,8 @@
 // if MARKING_STRESS_TEST is defined, we stress test the TryMark code
 // #define MARKING_STRESS_TEST
 
+#define DISABLE_BIGVAL_TRACKING
```
Comment (fingolfin) [inline src/julia_gc.c]: The correct way would be to uncomment the `#define` in line 43

File: src/julia_gc.c [hunk 2]
```diff
@@ -450,6 +452,10 @@ void InitMarkFuncBags(UInt type, TNumMarkFuncBags mark_func)
 
 static inline int JMark(void * obj)
 {
+    void *ty = jl_typeof(obj);
+    if (ty != datatype_mptr && ty != datatype_bag
+        && ty != datatype_largebag && ty != jl_weakref_type)
+        return 0;
```
Comment (fingolfin) [inline src/julia_gc.c]: Hm, so that's quite a bit different from the changes we discussed last. So, this is a hard coded list of types -- so we are making the hard assumption here that GAP objects will *never* directly reference other Julia objects. But that is not only fragile, it is also quite simply wrong, because JuliaInterface of course does so.

Hence, I don't understand how this can work at all.

I also don't quite understand why it is now sufficient to perform this check in `JMark` alone -- is it safe to increment `YoungRef` in `MarkBag` when dealing with invalid bags?

Comment (rbehrends) [reply src/julia_gc.c]: Hmm, I hadn't realized that JuliaInterface goes through `JMark()` also. This is something that needs to be fixed indeed. I understand the fragility concern, but this is a hot path, where we need to look up these types fast.

The `MarkBag()` case is safe. We have at this point already validated that `p` is a valid address. If on a freelist or if unmarked, then the lowest bit will be zero and we will not increment `YoungRef`.

I am still going to rewrite this for better readability (and, of course, to remove the `JMark()` problem.

File: src/julia_gc.c (hunk 1 above)
Comment (rbehrends) [reply src/julia_gc.c]: Will do.

File: src/julia_gc.c [hunk 3]
```diff
@@ -685,7 +701,7 @@ static void PostGCHook(int full)
 // is called by the Julia GC whenever it marks a GAP master pointer object)
 static uintptr_t JMarkMPtr(jl_ptls_t ptls, jl_value_t * obj)
 {
-    if (!*(void **)obj)
+    if (!*(void **)obj || !jl_typeis(obj, datatype_mptr))
```
Comment (fingolfin) [inline src/julia_gc.c]: But `obj` does not point at the start of the mptr, `BAG_HEADER((Bag)obj)` does.

File: src/julia_gc.c [hunk 4]
```diff
@@ -575,7 +591,7 @@ static void GapRootScanner(int full)
 {
     // mark our Julia module (this contains references to our custom data
     // types, which thus also will not be collected prematurely)
-    JMark(Module);
+    jl_gc_mark_queue_obj(JuliaTLS, (jl_value_t *) Module);
```
Comment (fingolfin) [inline src/julia_gc.c]: Why this change?

File: src/julia_gc.c [hunk 5]
```diff
@@ -513,7 +524,12 @@ static void TryMark(void * p)
 #endif
     }
     if (p2) {
-        JMark(p2);
+#ifdef DISABLE_BIGVAL_TRACKING
+        JMarkGapObjSafe(p2);
+#else
+        if (jl_typeis(p2, datatype_mptr))
+            JMark(p2);
```
Comment (fingolfin) [inline src/julia_gc.c]: Why not always use this? Why is the `DISABLE_BIGVAL_TRACKING` doing something else?

File: src/julia_gc.c (hunk 2 above)
Comment (fingolfin) [reply src/julia_gc.c]: Ah, I somehow overlooked that `MarkBag` already validates the pointer (which it indeed does).

File: src/julia_gc.c (hunk 4 above)
Comment (rbehrends) [reply src/julia_gc.c]: This allowed me to put validation logic in `JMark()` to verify that bad pointers indeed don't get traversed. I should actually put that in.

File: src/julia_gc.c (hunk 3 above)
Comment (rbehrends) [reply src/julia_gc.c]: Yes, that was actually an unnecessary test, too.

File: src/julia_gc.c (hunk 5 above)
Comment (rbehrends) [reply src/julia_gc.c]: The two cases were flipped around, or alternatively, an `#ifndef` should have been used. So the `DISABLE_BIGVAL_TRACKING` was actually more restrictive.

Comment (fingolfin) [reply src/julia_gc.c]: OK, but that doesn't really answer my question: why are the two cases doing something different? (Perhaps a code comment should explain it?)

File: src/julia_gc.c [hunk 6]
```diff
@@ -605,7 +634,7 @@ static void GapRootScanner(int full)
     for (Int i = 0; i < GlobalCount; i++) {
         Bag p = *GlobalAddr[i];
         if (IS_BAG_REF(p)) {
-            JMark(p);
+            JMarkGapObjSafe(p);
```
Comment (fingolfin) [inline src/julia_gc.c]: Why does this need to use `JMarkGapObjSafe`? It seems to be me that we even could use `jl_gc_mark_queue_obj` directly here, no?

File: src/julia_gc.c (hunk 5 above)
Comment (rbehrends) [reply src/julia_gc.c]: Okay, technically the `DISABLE_BIGVAL_TRACKING` should actually have a different name, namely that we only mark master pointers conservatively from the stack. Without the test, it would also mark small objects. I think the right thing to do here is to rename the define.

File: src/julia_gc.c (hunk 6 above)
Comment (rbehrends) [reply src/julia_gc.c]: It probably could. I'll have a look at that.

File: src/julia_gc.c [hunk 7]
```diff
@@ -37,10 +37,13 @@
 **  Various options controlling special features of the Julia GC code follow
 */
 
-// if DISABLE_BIGVAL_TRACKING is defined, we don't track the location of
-// large bags; this speeds up some things, but may expose bugs in GAP code
-// which incorrectly holds pointers into bags over a GC.
-// #define DISABLE_BIGVAL_TRACKING
+// if SCAN_STACK_FOR_MPTRS_ONLY is defined, stack scanning will only
+// look for references to master pointers, but not bags themselves.
+// This should be safe, as Gasman uses the same mechanism, but it is
```
Comment (fingolfin) [inline src/julia_gc.c]:
```suggestion
// This should be safe, as GASMAN uses the same mechanism, but it is
```

File: src/julia_gc.c [hunk 8]
```diff
@@ -37,10 +37,13 @@
...
+// at least theoretically possible that an optimizing compiler will
+// remove the last master pointer reference before the last reference
+// to the bag itself.
```
Comment (fingolfin) [inline src/julia_gc.c]: As discussed several times before: such an optimization would necessarily also affect GASMAN in the very same way! Except perhaps if the Julia GC was running concurrently to the thread GAP is running in; but then we might have a serious problem anyway.

File: src/julia_gc.c [hunk 9]
```diff
@@ -196,7 +205,7 @@ static inline void * align_ptr(void * p)
     return (void *)u;
 }
 
-#if !defined(DISABLE_BIGVAL_TRACKING)
+#if !defined(SCAN_STACK_FOR_MPTRS_ONLY)
```
Comment (fingolfin) [inline src/julia_gc.c]: ... seems it's a good thing I insisted on putting my `DISABLE_BIGVAL_TRACKING` mod in there... ;-)

File: src/julia_gc.c [hunk 10]
```diff
@@ -448,11 +457,51 @@ void InitMarkFuncBags(UInt type, TNumMarkFuncBags mark_func)
...
+#endif
     return jl_gc_mark_queue_obj(JuliaTLS, (jl_value_t *)obj);
 }
 
+void MarkJuliaObjSafe(void * obj)
+{
+    if (!obj)
+        return;
+    jl_value_t *ty = jl_typeof(obj);
+    if (jl_gc_internal_obj_base_ptr(ty) != ty)
```
Comment (fingolfin) [inline src/julia_gc.c]: So we call `jl_gc_internal_obj_base_ptr` on the type -- why don't we just call `jl_gc_internal_obj_base_ptr` on the object, i.e., `jl_gc_internal_obj_base_ptr(obj)`?

File: src/julia_gc.c [hunk 11]
```diff
@@ -448,11 +457,51 @@ void InitMarkFuncBags(UInt type, TNumMarkFuncBags mark_func)
...
+        return;
+    if (!jl_typeis(ty, jl_datatype_type))
+        return;
```
Comment (fingolfin) [inline src/julia_gc.c]: Please add comments that explain the checks above

File: src/julia_gc.c (hunk 10 above)
Comment (fingolfin) [reply src/julia_gc.c]: Same question of course applies to the debug code in `JMark`

File: src/julia_gc.c [hunk 12]
```diff
@@ -448,11 +457,51 @@ void InitMarkFuncBags(UInt type, TNumMarkFuncBags mark_func)
     TabMarkFuncBags[type] = mark_func;
 }
 
+static inline int JMarkGapObjSafe(void * obj)
+{
+    // only traverse objects internally used by GAP
+    void *ty = jl_typeof(obj);
+    if (ty != datatype_mptr && ty != datatype_bag
+        && ty != datatype_largebag && ty != jl_weakref_type)
```
Comment (fingolfin) [inline src/julia_gc.c]: There are three calls to `JMarkGapObjSafe`. Two are in `MarkBag` -- but for that caller, we only need to accept `datatype_mptr` and `jl_weakref_type`. And if we added a custom `MarkBagWeakly` function (to mirror the name used in GASMAN) which only accepts `jl_weakref_type`, we could reduce the check to `jl_typeis(obj, datatype_mptr)` which is more efficient.

And `MarkBagWeakly` could potentially be considerably simpler than `MarkBag`, as (a) it is very rarely used, and (b) we control where it is called from tightly, and so we'd not have to validate as much, and could skipping the `MarkCache`. So this might suffice:
```C
void MarkBagWeakly(Bag bag)
{
    if (bag == 0)
        return;
    jl_value_t * p = (jl_value_t *)bag;
    GAP_ASSERT(jl_typeis(p, jl_weakref_type));
    if (jl_gc_mark_queue_obj(JuliaTLS, p))
        YoungRef++;
}
```


Anyway, that leaves the third call to `JMarkGapObjSafe`, which is in `TryMark`. Note that in the `SCAN_STACK_FOR_MPTRS_ONLY` case, it also only checks `jl_typeis(obj, datatype_mptr)`.

So, let's change `JMarkGapObjSafe` to only check for `datatype_mptr`, and then replace the one remaining call to it with a copy of the code that's now here. And then we could rename this function to something more specific... I want to say `JMarkMPtr`, but we already use that name (which, BTW, I find rather confusing -- `JMarkMPtr` is a Julia marking function, while `JMark` is a helper function for use in lots of places

File: src/julia_gc.c [hunk 13]
```diff
@@ -687,7 +741,11 @@ static uintptr_t JMarkMPtr(jl_ptls_t ptls, jl_value_t * obj)
 {
     if (!*(void **)obj)
         return 0;
-    if (JMark(BAG_HEADER((Bag)obj)))
+    void *header = BAG_HEADER((Bag) obj);
```
Comment (fingolfin) [inline src/julia_gc.c]:
```suggestion
    void *header = BAG_HEADER((Bag)obj);
```

File: src/julia_gc.c [hunk 14]
```diff
@@ -448,11 +457,51 @@ void InitMarkFuncBags(UInt type, TNumMarkFuncBags mark_func)
...
+    jl_value_t *ty = jl_typeof(obj);
+    if (jl_gc_internal_obj_base_ptr(ty) != ty)
+        abort();
+    if (!jl_typeis(ty, jl_datatype_type))
+        abort();
+#endif
     return jl_gc_mark_queue_obj(JuliaTLS, (jl_value_t *)obj);
 }
 
+void MarkJuliaObjSafe(void * obj)
```
Comment (sebasguts) [inline src/julia_gc.c]: Since we need those functions (`MarkJuliaObjSafe` and `MarkJuliaObj`) in JuliaInterface, could you expose them in a header?

And, out of curiosity, is there a reason that the argument type is `void*` and not `jl_value_t*`?

Comment (fingolfin) [reply src/julia_gc.c]: Because `void *` is much easier to expose in a header (doesn't force all GAP code to include Julia headers)

Comment (sebasguts) [reply src/julia_gc.c]: Makes sense, thank you for the clarification.

File: src/julia_gc.c (hunk 10 above)
Comment (rbehrends) [reply src/julia_gc.c]: At this point we already know that `obj` points to a proper object, which may be either a pool or large object. What the check does is verifying that it actually is a proper Julia object and has not been freed; if freed, the type tag would be a freelist pointer and fail that check.

File: src/julia_gc.c (hunk 12 above)
Comment (rbehrends) [reply src/julia_gc.c]: In `TryMark()`, `JMarkGapObjSafe()` is not called if `SCAN_STACK_FOR_MPTRS_ONLY` is enabled. Otherwise, it's used in lieu of the MPtr type check and needs to cover all cases.

I did not specialize the type check in `JMarkGapObjSafe()` further for other use cases, because it will succeed on the first hit for almost all cases with `SCAN_STACK_FOR_MPTRS_ONLY` enabled and is probably an unnecessary optimization.

File: src/julia_gc.c (hunk 13 above)
Comment (rbehrends) [reply src/julia_gc.c]: Will do.

File: src/julia_gc.c (hunk 14 above)
Comment (rbehrends) [reply src/julia_gc.c]: Yes, we can definitely put that in a header.

File: src/julia_gc.c (hunk 12 above)
Comment (rbehrends) [reply src/julia_gc.c]: I am on board with renaming the functions, though. That thought had occurred to me, too.

File: src/julia_gc.c (hunk 8 above)
Comment (fingolfin) [reply src/julia_gc.c]: To be clear: With that remark I meant to say that I think your comment is misleading, and should be adjusted. Either removed, or qualified.

File: src/julia_gc.c [hunk 15]
```diff
@@ -687,7 +747,11 @@ static uintptr_t JMarkMPtr(jl_ptls_t ptls, jl_value_t * obj)
 {
     if (!*(void **)obj)
         return 0;
-    if (JMark(BAG_HEADER((Bag)obj)))
+    void *header = BAG_HEADER((Bag)obj);
+    void *ty = jl_typeof(header);
+    if (ty != datatype_bag && ty != datatype_largebag)
```
Comment (fingolfin) [inline src/julia_gc.c]: Please add a brief comment here which explains why this check is performed.

File: src/julia_gc.c [hunk 16]
```diff
@@ -575,7 +636,7 @@ static void GapRootScanner(int full)
 {
     // mark our Julia module (this contains references to our custom data
     // types, which thus also will not be collected prematurely)
-    JMark(Module);
+    jl_gc_mark_queue_obj(JuliaTLS, (jl_value_t *) Module);
```
Comment (fingolfin) [inline src/julia_gc.c]: Please add a brief comment here which explains why this does not call `JMark` resp. `MarkJuliaObj` (probably something like this: 
```suggestion
    // We call jl_gc_mark_queue_obj directly here instead of one of our wrappers
    // because we know that <Module> is a valid Julia object here; so we can and
    // should bypass any validation checks.
    jl_gc_mark_queue_obj(JuliaTLS, (jl_value_t *)Module);
```
Although this does not yet explain why `YoungRef` is not relevant here, which should also be stated.

File: src/julia_gc.h [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+**
+**  This file declares functions of the Julia GC interface.
+*/
+
+#ifndef GAP_JULIA_GC_H
+#define GAP_JULIA_GC_H
+
+/****************************************************************************
+**
+*F  MarkJuliaObj(<obj>)  . . . . . . . . . . . . . . . . mark Julia object
```
Comment (fingolfin) [inline src/julia_gc.h]:
```suggestion
*F  MarkJuliaObj(<obj>) . . . . . . . . . . . . . . . . . . mark Julia object
```

File: src/julia_gc.h [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+**
+**  'MarkJuliaObjSafe' marks a Julia object; the argument can be NULL.
```
Comment (fingolfin) [inline src/julia_gc.h]:
```suggestion
**  'MarkJuliaObj' marks a Julia object; the argument may be NULL. No further
**  checks are performed. If <obj> is not a Julia object, a crash may result.
```

File: src/julia_gc.h [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+*/
+
+void MarkJuliaObj(void * obj);
+
+/****************************************************************************
+**
+*F  MarkJuliaObjSafe(<obj>)  . . . . . . . . . . . . . . mark Julia object
+**
+**  'MarkJuliaObjSafe' marks a Julia object, verifying that it is a valid
+**  object.
```
Comment (fingolfin) [inline src/julia_gc.h]:
```suggestion
```

File: src/julia_gc.h [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+**  'MarkJuliaObjSafe' marks a Julia object; the argument can be NULL.
+*/
+
+void MarkJuliaObj(void * obj);
+
+/****************************************************************************
+**
+*F  MarkJuliaObjSafe(<obj>)  . . . . . . . . . . . . . . mark Julia object
+**
+**  'MarkJuliaObjSafe' marks a Julia object, verifying that it is a valid
```
Comment (fingolfin) [inline src/julia_gc.h]:
```suggestion
**  'MarkJuliaObjSafe' marks a Julia object; the argument may be NULL.
**  Extra validation steps are performed to determine whether <obj> is a valid
**  Julia object pointer; if not, it is silently ignored.
```

File: src/julia_gc.h (hunk 19 above)
Comment (rbehrends) [reply src/julia_gc.h]: This does not seem to be correct?

File: src/julia_gc.c [hunk 21]
```diff
@@ -30,17 +30,20 @@
...
 **  Various options controlling special features of the Julia GC code follow
 */
 
-// if DISABLE_BIGVAL_TRACKING is defined, we don't track the location of
-// large bags; this speeds up some things, but may expose bugs in GAP code
-// which incorrectly holds pointers into bags over a GC.
-// #define DISABLE_BIGVAL_TRACKING
+// if SCAN_STACK_FOR_MPTRS_ONLY is defined, stack scanning will only
+// look for references to master pointers, but not bags themselves.
+// This should be safe, as GASMAN uses the same mechanism and is therefore
```
Comment (fingolfin) [inline src/julia_gc.c]: Wouldn't this be more accurate?
```suggestion
// This should be safe, as GASMAN uses the same mechanism. It is also faster
// and avoids certain complicated issues that can lead to crashes, and is therefore
```

File: src/julia_gc.h [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+**  'MarkJuliaObjSafe' marks a Julia object; the argument may be NULL.
+**  Extra validation steps are performed to determine whether <obj> is
+**  a valid Julia object. If not, it is silently ignored.
+*/
+
+void MarkJuliaObjSafe(void * obj);
+
+/****************************************************************************
+**
+*F  MarkJuliaWeakRef(<ref>)  . . . . . . . . . .  mark Julia weak reference
```
Comment (fingolfin) [inline src/julia_gc.h]:
```suggestion
*F  MarkJuliaWeakRef(<ref>) . . . . . . . . . . . . mark Julia weak reference
```

File: src/julia_gc.h [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+**
+**  This file declares functions of the Julia GC interface.
+*/
+
+#ifndef GAP_JULIA_GC_H
+#define GAP_JULIA_GC_H
+
+/****************************************************************************
+**
+*F  MarkJuliaObj(<obj>)  . . . . . . . . . . . . . . . . . mark Julia object
```
Comment (fingolfin) [inline src/julia_gc.h]:
```suggestion
*F  MarkJuliaObj(<obj>) . . . . . . . . . . . . . . . . . . mark Julia object
```

File: src/julia_gc.h [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+**  'MarkJuliaObjSafe' marks a Julia object; the argument can be NULL. No
+**  further checks are performed. If <obj> is not a Julia object, a crash
+**  may result.
+*/
+
+void MarkJuliaObj(void * obj);
+
+/****************************************************************************
+**
+*F  MarkJuliaObjSafe(<obj>)  . . . . . . . . . . . . . . . mark Julia object
```
Comment (fingolfin) [inline src/julia_gc.h]:
```suggestion
*F  MarkJuliaObjSafe(<obj>) . . . . . . . . . . . . . . . . mark Julia object
```

File: src/julia_gc.c [hunk 25]
```diff
@@ -448,11 +457,55 @@ void InitMarkFuncBags(UInt type, TNumMarkFuncBags mark_func)
     TabMarkFuncBags[type] = mark_func;
 }
 
+static inline int JMarkTyped(void * obj, jl_datatype_t * ty)
```
Comment (fingolfin) [inline src/julia_gc.c]: OK, that seems fine!

But, if we use this name with `JMark` prefix here now, please let's at least rename `JMarkMPtr` and `JMarkBag` -- e.g. to `JuliaMarkFunctionForMPtrs` and `JuliaMarkFunctionForBags`.

File: src/julia_gc.c [hunk 26]
```diff
@@ -448,11 +457,55 @@ void InitMarkFuncBags(UInt type, TNumMarkFuncBags mark_func)
...
+    if (jl_gc_mark_queue_obj(JuliaTLS, (jl_value_t *)obj))
+        YoungRef++;
+}
+
+
+void MarkJuliaObj(void * obj)
+{
+    if (!obj)
+        return;
+    if (jl_gc_mark_queue_obj(JuliaTLS, (jl_value_t *)obj))
```
Comment (fingolfin) [inline src/julia_gc.c]: Should we perhaps call `JMark` here so that when `VALIDATE_MARKING` is enabled, we validate here, too?

File: src/julia_gc.c (hunk 25 above)
Comment (rbehrends) [reply src/julia_gc.c]: I picked `MPtrMarkFunc` and `BagMarkFunc`, respectively. Those should be both clear and distinct enough.

File: src/julia_gc.c (hunk 26 above)
Comment (rbehrends) [reply src/julia_gc.c]: Good point, will put that in.
